File: lib/ts/TransactionWriter.ts
import { TraceProject, TraceTransactionLog } from '../../models/ts/Tracer_pb';

export abstract class TransactionWriter {
    private savedTransactionLogPartions: number[] = new Array<number>();

    constructor() {
    }

    public Reset(): void {
        this.savedTransactionLogPartions = new Array<number>();
    }

    public async SaveTransactionLog(transactionLog: TraceTransactionLog, projectId: string): Promise<boolean> {
        const buffer = transactionLog.serializeBinary();
        console.log(`saving ${JSON.stringify(transactionLog.toObject())}`);
        return await this.WriteTransactionLog(transactionLog, buffer, projectId);
    }

    public async SaveTransactionLogs(transactionLogs: TraceTransactionLog[], projectId: string): Promise<boolean> {
        let success = true;
        for (const transactionLog of transactionLogs) {
            if (this.savedTransactionLogPartions.indexOf(transactionLog.getPartition(), 0) !== -1) {
                continue;
            }
            const saveResult = await this.SaveTransactionLog(transactionLog, projectId);
            if (saveResult) {
                this.savedTransactionLogPartions.push(transactionLog.getPartition());
            }

            success = success && saveResult;
        }

        return success;
    }

    // tslint:disable-next-line: max-line-length
    protected async abstract WriteTransactionLog(transactionLog: TraceTransactionLog, data: Uint8Array, projectId: string): Promise<boolean>;
}

File: lib/cs/TransactionWriter.cs
using System.Collections.Generic;
using System.IO;
using Google.Protobuf;

namespace Tracer
{

    public abstract class ITransactionWriter
    {
        public TraceProject Project { get; private set; }

        //For loading existing transactions
        //Logs should only be after the offset
        public ITransactionWriter(TraceProject project)
        {
            Project = project;
        }

        public void SaveProject()
        {
            using (var stream = new MemoryStream())
            {
                Project.WriteTo(stream);
                CreateProject(stream);
            }
        }

        public void SaveTransactionLog(TraceTransactionLog transactionLog)
        {
            using (var stream = new MemoryStream())
            {
                transactionLog.WriteTo(stream);
                WriteTransactionLog(transactionLog, stream);
            }
        }

        public void SaveTransactionLogs(List<TraceTransactionLog> transactionLogs)
        {
            foreach (var transactionLog in transactionLogs)
            {
                using (var stream = new MemoryStream())
                {
                    transactionLog.WriteTo(stream);
                    WriteTransactionLog(transactionLog, stream);
                }
            }
        }

        protected abstract void CreateProject(Stream data);
        protected abstract void WriteTransactionLog(TraceTransactionLog transactionLog, Stream data);
    }
}
